### DISCUSSION TITLE
mongodb replica set in gitlab-ci

### DISCUSSION THREAD

**Discussion by SimoneBergonzi** [mia-platform/community#135] [Q&A] [answered]
Hi,

I'm working on a service which needs to use mongo change streams and I'm having problems setting up a mongo replica set for the test stage of the gitlab-ci.

I tried with these ready-to-use docker images:
- https://hub.docker.com/r/bitnami/mongodb/
- https://registry.hub.docker.com/r/candis/mongo-replica-set
but still the service won't connect to the mongo cluster.

Is there anybody who worked with mongo change streams or anything that requires a replica set that could help me?
Thanks

**Comment by davidebianchi** [comment, marked as answer]
Hi @SimoneBergonzi! Locally it is possible to use something like:

```sh
docker run -d -p 27017:27017 --name mongo_test mongo mongod --replSet testRepl 
docker exec -i -t mongo_test mongo localhost --eval "rs.initiate()"
```

to start mongo with replicaset. 
Some interesting links you can use to investigate for a possible solution on gitlab CI: [link 1](https://gitlab.com/gitlab-org/gitlab/-/issues/333636#note_601476144), [link 2](https://www.mongodb.com/community/forums/t/start-single-node-replica-set-in-gitlab-ci/125970).

**Reply by frengooo** [reply]
Locally works using `rs.initiate()`, on the CI it's not working.
Is there a way, from the CI, to run a script directly inside a docker?
e.g. to run `docker exec -it mongo mongosh --eval 'rs.initiate()'`

**Reply by davidebianchi** [reply]
Have you seen the link I shared, like [this](https://www.mongodb.com/community/forums/t/start-single-node-replica-set-in-gitlab-ci/125970/17)?

**Reply by frengooo** [reply]
Yes, we tried all this approaches 😂 
We tried:
To initiate the rs from the start command directly, in this way (in local it works):
```
services:
    - name: mongo:5
      alias: mongo
      command: ["/bin/sh", "-c", "mongod --logpath /dev/null --bind_ip_all --replSet 'rs0' && mongosh --eval 'rs.initiate()'"]
```
&rarr; but it doesn't work, the tests are unable to reach the cluster

To install `mongo` and run `rs.initiate()` in before scripts:
```
before_script:
    - wget -qO - https://www.mongodb.org/static/pgp/server-5.0.asc |  apt-key add -
    - echo "deb http://repo.mongodb.org/apt/debian buster/mongodb-org/5.0 main" |  tee /etc/apt/sources.list.d/mongodb-org-5.0.list
    - apt-get update
    - apt-get install -y mongodb-org
    - mongo --host mongo --eval 'rs.initiate()'
```
&rarr; but it doesn't work because tests run in `docker in docker` and we can't install stuffs

**Reply by davidebianchi** [reply]
After the first approach, what is the error given by the pipeline?

**Reply by frengooo** [reply]
After the first approach the tests failed for timeout during the connection to mongo.

Anyway, we fixed using the `before_script` attached below. The problem wasn't `dind`, but the node image we were using.
Using the `alpine` version we couldn't run apt-get and other commands, without the alpine it works well.

Btw it's not the cleanest solution ever 😅

**Reply by davidebianchi** [reply]
Yes, I agree with you. So now it works? 
With alpine you should use `apk`. 

A possible solution could be to create a custom docker image which already set the mongo replicas up?

**Reply by frengooo** [reply]
Now it's working 😄
I don't really like the "custom docker image" approach, but we probably will try it or some other solution.

Thanks guys 😄
